### PR TITLE
IS-17846 add reusable shopify webhook docs

### PR DESCRIPTION
**Pull request by parhambarazesh**

### CODE REVIEW

**Comment by parhambarazesh** [comment]
Screenshot from 2024-10-02 11-50-23
Screenshot from 2024-10-02 11-51-10
Screenshot from 2024-10-02 11-51-15
Screenshot from 2024-10-02 11-51-26

**Review by erik-leven — COMMENTED** [review]
Include `partial webhook events` and `shared webhook events` in `receiving webhooks`.
Add example of full pipe setup (incl. even2 for shared event and collect pipe with merge_datasets source)
Clean up examples where settings are system specific
Partial webhook evens are really just a parameterized input. Should atleast mention this

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 1]
```diff
@@ -1,3 +1,181 @@
...
+* some systems use a default secret key, e.g. Shopify uses the ``client_secret`` to sign the webhook payload. In this case, the event pipe must have the correct validation expression that uses the ``client_secret`` and ``-registerwebhoook`` pipe don't send any secret key.
+
+Example: validation expression
+------------------------------
+
+::
+
+   "validation_expression": "{% if request_headers['X-Shopify-Hmac-SHA256'] == b64encode(hmacsha256digest(secret('oauth_client_secret'), request_body)) %}{% else %}FAIL!{% endif %}"
+
+* When testing the webhooks with Sesampy on dev nodes, this is not automatically handled and you may need to manually put the ``client_secret`` value into the validation expression.
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:22]: link to sesampy in docs

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 2]
```diff
@@ -1,3 +1,181 @@
...
+      ],
+      "type": "pipe"
+    }
+
+A full connector example of ``-event2`` pipe can be found in the `Hubspot connector's playground branch <https://github.com/sesam-io/hubspot-connector/blob/playground>`__ in the `company template <https://github.com/sesam-io/hubspot-connector/blob/playground/templates/company.json>`__.
+
+Shared webhook events
+---------------------
+
+If multitenant, receive all in one dataset and use subset expression with tenant id to filter out relevant events into tenant pipes. The tenant pipe should be called ``-event``.
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:127]: I don't think we have used the word multitenant for this before? I think it's easier if we just describe the example, i.e. "if all events regardless of datatype is received by the same pipe". Also after you use the word "tenant", but its all the same tenant really isn't it? Could we use "the datatype specific pipe" instead?

Comment (erik-leven) [reply hub/documentation/connectors/patterns/receive-webhook-events.rst:127]: Also, in this case you have 2 pipes, one to receive all webhooks and one downstream to receive the webhooks for a specific datatype. This should be clearer in the text, and potentially with an example of the subset expression

Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:127]: Link to subset expression?

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 3]
```diff
@@ -1,3 +1,181 @@
...
+If when webhook events are shared, we need to separate them into datatype specific pipes before merging.
+
+If we have webhooks they need to combine with full scan pipes in the collect pipe.
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:130]: Maybe a new subsection here called "combining event and non event data"

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 4]
```diff
@@ -1,3 +1,181 @@
...
+
+Example: -collect pipe (webhook)
+--------------------------------
+
+::
+
+  {
+      "_id": "<system>-<datatype>-collect",
+      "exclude_completeness": [
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:139]: remove exclude completeness. Not important for webhooks

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 5]
```diff
@@ -1,3 +1,181 @@
...
+        "<system>-<datatype>-share",
+        "<system>-<datatype>-event2",
+        "<system>-<datatype>-collect"
+      ],
+      "namespaced_identifiers": false,
+      "pump": {
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:145]: remove pump

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 6]
```diff
@@ -1,3 +1,181 @@
...
+      },
+      "source": {
+        "datasets": [
+          "<system>-<datatype>-all",
+          "<system>-<datatype>-event2"
+        ],
+        "type": "merge_datasets"
+      },
+      "transform": [
+        {
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:156]: remove this lookup transform

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 7]
```diff
@@ -1,3 +1,181 @@
...
+            "operation_lookup_delete": "<datatype>-lookup",
+            "primary_key": "id",
+            "rest_system": "<system>",
+            "share_dataset": "<system>-<datatype>-share"
+          },
+          "template": "transform-collect-rest",
+          "type": "template"
+        },
+        {
+          "rules": {
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:167]: remove this transform (not needed when only doing ["copy", "*"]

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 8]
```diff
@@ -1,3 +1,226 @@
...
+    }
+
+A full connector example of ``-event2`` pipe can be found in the `Hubspot connector's playground branch <https://github.com/sesam-io/hubspot-connector/blob/playground>`__ in the `company template <https://github.com/sesam-io/hubspot-connector/blob/playground/templates/company.json>`__.
+
+Shared webhook events
+---------------------
+
+If webhook events are shared, we need to separate them into datatype specific pipes before merging. When all events, regardless of their datatype, are received by a single pipe, we can filter out relevant events using a subset expression. This allows us to route events of a specific datatype to another pipe.
+
+* The -all pipe gathers all incoming webhook events.
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:129]: there is no -all pipe here. Only a general webhook even pipe

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 9]
```diff
@@ -1,3 +1,226 @@
...
+        "datasets": [
+          "<system>-<datatype>-all",
+          "<system>-<datatype>-event2"
+        ],
+        "type": "merge_datasets"
+      },
+      "transform": [
+        {
+          "rules": {
+            "default": [
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:213]: remove transform is only doing ["copy", "*"]

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 10]
```diff
@@ -1,3 +1,226 @@
...
+      "deletion_tracking": false,
+      "set_initial_offset": "onload"
+    },
+    "source": {
+      "dataset": "<hubspot_webhook_dataset>",
+      "subset": [
+        "eq",
+        [
+          "and",
+          [
```
Comment (erik-leven) [inline hub/documentation/connectors/patterns/receive-webhook-events.rst:150]: is this check required? Looks very Talk dependent?

File: hub/documentation/connectors/patterns/receive-webhook-events.rst [hunk 11]
```diff
@@ -1,3 +1,226 @@
...
+              "list",
+              "_S.<property>"
+            ]
+          ]
+        ],
+        true
+      ],
+      "type": "dataset"
+    },
+    "transform": [
```
[the remark this answers is not in the document]
Comment (parhambarazesh) [reply hub/documentation/connectors/patterns/receive-webhook-events.rst]: done

File: hub/documentation/connectors/patterns/receive-webhook-events.rst (hunk 8 above)
Comment (parhambarazesh) [reply hub/documentation/connectors/patterns/receive-webhook-events.rst]: done

File: hub/documentation/connectors/patterns/receive-webhook-events.rst (hunk 3 above)
Comment (parhambarazesh) [reply hub/documentation/connectors/patterns/receive-webhook-events.rst]: done

File: hub/documentation/connectors/patterns/receive-webhook-events.rst (hunk 2 above)
Comment (parhambarazesh) [reply hub/documentation/connectors/patterns/receive-webhook-events.rst]: done

Comment (parhambarazesh) [reply hub/documentation/connectors/patterns/receive-webhook-events.rst]: done

**Review by erik-leven — APPROVED** [review]

**Review by hglad — APPROVED** [review]

**Pull request merged**